feat: configure standardResources via controller configmap

inishchith opened the pull request:
Better way to configure the memory and CPU resource requests


closes #1487

Sample configmap

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: numaflow-controller-config
data:
  controller-config.yaml: |+
    defaults:
      containerResources: |
        limits:
          memory: "256Mi"
          cpu: "200m"
        requests:
          memory: "128Mi"
          cpu: "100m"
```

<!--

Before you push your changes:

* Run `make pre-push -B` to fix codegen and lint problems.

Then, you MUST:

* Sign-off your commit (otherwise the DCO check will fail).
* Use [a conventional commit message](https://www.conventionalcommits.org/en/v1.0.0/) (otherwise the commit message check will fail).

If you did not do this, reset all your commit and replace them with a single commit:

```
git reset HEAD~1 ;# change 1 to how many commits you made
git commit --signoff -m 'feat: my feat. Fixes #1234'
```

When creating your PR: 

* Make sure that "Fixes #" or "Closes #" is in both the PR title (for release notes) and description (to automatically link and close the issue).
* Say how you tested your changes. If you changed the UI, attach screenshots.
* Set your PR as a draft initially.
* Your PR needs to pass the required checks before it can be approved. 
* Once required tests have passed, mark your PR "Ready for review".

If changes were requested, once you've made them, you MUST dismiss the review to get it reviewed again.

-->

vigith commented:
Is standardResources only for UDF containers?

whynowy commented on pkg/reconciler/config.go:
```diff
@@ -21,13 +21,21 @@ import (
 // controller manager.
 type GlobalConfig struct {
-	ISBSvc *ISBSvcConfig `json:"isbsvc"`
+	StandardResources *ResourceConfig `json:"standardResources"`
```
```
data:
  controller-config.yaml: |+
    defaults:
      containerResources:
          limits:
            cpu: xxxx
            memory: xxxx
          requests:
            cpu: xxx
            memory: xxxx
    isbsvc:
      redis:
...
```

This way we may add more `defaults` in the future.

Also, use `*apiv1.ResourceRequirements` directly.

inishchith replied:
Agree with additional defaults in the future. ResourceList is a Map which makes the valid yaml as following, if I were to use `*apiv1.ResourceRequirements` directly.

```yaml
defaults:
  containerResources:
    limits:
      memory:
        value: xxxx
      cpu:
        value: xxxx
    requests:
      memory:
        value: xxxx
      cpu:
        value: xxxx
```

checking if there could be a simpler way to handle this while opening up scope for more defaults.

inishchith replied:
Viper uses the mapstructure package for decoding, which doesn't support resource.Quantity out of the box.

inishchith replied:
updated to cover defaults

whynowy replied:
If Viper does not support, make it a string, and then we parse it.

whynowy replied:
```
data:
  controller-config.yaml: |+
    defaults:
      containerResources: |
          limits:
            cpu: xxxx
            memory: xxxx
          requests:
            cpu: xxx
            memory: xxxx
    isbsvc:
      redis:

whynowy replied:
Actually I didn't get it, Viper does not support customized structs? we are already doing that, right?

inishchith replied:
> Viper does not support customized structs?
No, it does not parse `resource.Quantity` i.e the value of the map object (under limits/resource) correctly - assigns them 0. Which is due to the underlying mapstructure.

inishchith replied:
> If Viper does not support, make it a string, and then we parse it.

Better. updated!

whynowy commented on pkg/reconciler/config.go:
```diff
@@ -69,6 +77,27 @@ type JetStreamVersion struct {
 	StartCommand         string `json:"startCommand"`
 }
 
+func (g *GlobalConfig) GetStandardResources() corev1.ResourceRequirements {
```
`GetDefaultContainerResources()`, and it always return a new instance, so that the invokers do not need to deepCopy.

inishchith replied:
updated

whynowy commented on pkg/apis/numaflow/v1alpha1/get_spec_req.go:
```diff
@@ -21,21 +21,22 @@ import (
+	ConfConfigMapName         string                      `protobuf:"bytes,13,opt,name=confConfigMapName"`
+	ScriptsConfigMapName      string                      `protobuf:"bytes,14,opt,name=scriptsConfigMapName"`
+	HealthConfigMapName       string                      `protobuf:"bytes,15,opt,name=healthConfigMapName"`
+	StandardResources         corev1.ResourceRequirements `protobuf:"bytes,16,opt,name=standardResources"`
```
Use `DefaultResources`?

whynowy commented on pkg/apis/numaflow/v1alpha1/get_spec_req.go:
```diff
@@ -45,35 +46,38 @@ type GetRedisServiceSpecReq struct {
+	PullPolicy          corev1.PullPolicy           `protobuf:"bytes,3,opt,name=pullPolicy,casttype=k8s.io/api/core/v1.PullPolicy"`
+	Env                 []corev1.EnvVar             `protobuf:"bytes,4,rep,name=env"`
+	SideInputsStoreName string                      `protobuf:"bytes,5,opt,name=sideInputsStoreName"`
+	StandardResources   corev1.ResourceRequirements `protobuf:"bytes,6,opt,name=standardResources"`
```
Use `DefaultResources`?

whynowy commented on pkg/apis/numaflow/v1alpha1/get_spec_req.go:
```diff
@@ -45,35 +46,38 @@ type GetRedisServiceSpecReq struct {
+	Image             string                      `protobuf:"bytes,2,opt,name=image"`
+	PullPolicy        corev1.PullPolicy           `protobuf:"bytes,3,opt,name=pullPolicy,casttype=k8s.io/api/core/v1.PullPolicy"`
+	Env               []corev1.EnvVar             `protobuf:"bytes,4,rep,name=env"`
+	StandardResources corev1.ResourceRequirements `protobuf:"bytes,5,opt,name=standardResources"`
```
Same here.

whynowy commented on pkg/apis/numaflow/v1alpha1/get_spec_req.go:
```diff
@@ -45,35 +46,38 @@ type GetRedisServiceSpecReq struct {
+	ConfigMapName              string                      `protobuf:"bytes,12,opt,name=configMapName"`
+	PvcNameIfNeeded            string                      `protobuf:"bytes,13,opt,name=pvcNameIfNeeded"`
+	StartCommand               string                      `protobuf:"bytes,14,opt,name=startCommand"`
+	StandardResources          corev1.ResourceRequirements `protobuf:"bytes,15,opt,name=standardResources"`
```
Same here.

whynowy commented on pkg/apis/numaflow/v1alpha1/get_spec_req.go:
```diff
@@ -85,8 +89,9 @@ type GetJetStreamServiceSpecReq struct {
+	Image             string                      `protobuf:"bytes,2,opt,name=image"`
+	PullPolicy        corev1.PullPolicy           `protobuf:"bytes,3,opt,name=pullPolicy,casttype=k8s.io/api/core/v1.PullPolicy"`
+	Env               []corev1.EnvVar             `protobuf:"bytes,4,rep,name=env"`
+	StandardResources corev1.ResourceRequirements `protobuf:"bytes,5,opt,name=standardResources"`
```
And this.

inishchith commented:
> Is standardResources only for UDF containers?

@vigith - Not just UDF. All init and main containers (vertex, deamon redis_buffer_service,)

whynowy commented:
@inishchith - It looks good to me, can you please:

1. Add the new config to https://github.com/numaproj/numaflow/blob/main/config/base/controller-manager/numaflow-controller-config.yaml#L7, use the default resource settings;
2. Add some doc to https://github.com/numaproj/numaflow/blob/main/docs/operations/controller-configmap.md, also add the new config to https://github.com/numaproj/numaflow/blob/main/docs/operations/numaflow-controller-config.yaml.
3. Appreciate if you could add a config_test.go to cover the unit tests (unit tests for config.go are missing, you can add the one for the new function you added, but always welcome if you could cover more).

inishchith commented:
@whynowy Thanks!

I have updated the docs and config.
I have also added minimal tests that covers the newly introduced method. Covering the config file might require refactoring some `LoadConfig`, which I think might diverge from the goal of this change list. Let me know your thoughts.

whynowy commented on pkg/reconciler/config_test.go:
```diff
@@ -0,0 +1,76 @@
+// add test for config.go
```
Remove these.

whynowy commented on pkg/reconciler/config.go:
```diff
@@ -69,6 +77,29 @@ type JetStreamVersion struct {
+	var resourceConfig corev1.ResourceRequirements
+	err := yaml.Unmarshal([]byte(g.Defaults.ContainerResources), &resourceConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed unmarshal configuration file. %w", err)
```
`panic(ftm.Errorf("failed to unmarshal default container resources, %w", err))`

In this way there's no need to return an error but a struct only, that will make the invokers easier - letting the reconciler capture the error and keep retrying does not make sense.

whynowy commented on docs/operations/controller-configmap.md:
````diff
@@ -15,11 +15,54 @@ metadata:
+          cpu: "100m"
+```
+
+Similarly, you can set the default container limits resources for steps across all the pipelines:
````
This section has no difference from above, can be removed?

inishchith replied:
it has a difference, it explains not setting request. let me know if need to remove this

whynowy replied:
Why not setting `requests` it will default to limits?

inishchith replied:
As the ResourceRequirements are directly sourced to k8s API that we use ([refer](https://github.com/kubernetes/api/blob/master/core/v1/types.go#L2553)).

inishchith replied:
removed it!

whynowy commented on pkg/reconciler/config.go:
```diff
@@ -69,6 +77,29 @@ type JetStreamVersion struct {
 	StartCommand         string `json:"startCommand"`
 }
 
+func (g *GlobalConfig) GetDefaultContainerResources() *corev1.ResourceRequirements {
```
Now you can return the struct directly.

inishchith replied:
thanks for pointing out. fixed.

whynowy reviewed (approved):
LGTM, thanks!
